pkg/execution/svccatadmission/internaldns.go

```diff
@@ -0,0 +1,147 @@
+	ext_v1b1 "k8s.io/api/extensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// InternalDNSAdmitFunc checks existing DNS alias ownership via micros server API, for both InternalDNS Services and Ingress Resources
+func InternalDNSAdmitFunc(ctx context.Context, microsServerClient microsServerClient, scClient serviceCentralClient, admissionReview admissionv1beta1.AdmissionReview) (*admissionv1beta1.AdmissionResponse, error) {
+	logger := logz.RetrieveLoggerFromContext(ctx).Sugar()
+	admissionRequest := admissionReview.Request
+
+	// populating a chanel of domains to check owner
+	var domainsToCheck chan string
+	switch admissionRequest.Resource {
+	case k8s.IngressGVR:
+		ingress := ext_v1b1.Ingress{}
+		if err := json.Unmarshal(admissionRequest.Object.Raw, &ingress); err != nil {
+			return nil, errors.Errorf("error parsing Ingress resource")
+		}
+		if len(ingress.Spec.Rules) == 0 {
+			return nil, errors.Errorf("cannot process Ingress with empty rules list")
+		}
+		domainsToCheck = make(chan string, len(ingress.Spec.Rules))
+		for _, ingressRule := range ingress.Spec.Rules {
+			domainsToCheck <- ingressRule.Host
+		}
+	case k8s.ServiceInstanceGVR:
+		serviceInstance := sc_v1b1.ServiceInstance{}
+		if err := json.Unmarshal(admissionRequest.Object.Raw, &serviceInstance); err != nil {
+			return nil, errors.Wrap(err, "error parsing ServiceInstance")
+		}
+		var internalDNSSpec apiinternaldns.Spec
+		if err := json.Unmarshal(serviceInstance.Spec.Parameters.Raw, &internalDNSSpec); err != nil {
+			return nil, errors.Wrap(err, "error parsing InternalDNS spec")
+		}
+		if len(internalDNSSpec.Aliases) == 0 {
+			return nil, errors.Errorf("cannot process InternalDNS with empty aliases list")
+		}
+		domainsToCheck = make(chan string, len(internalDNSSpec.Aliases))
+		for _, alias := range internalDNSSpec.Aliases {
+			domainsToCheck <- alias.Name
+		}
+	default:
+		return nil, errors.Errorf("unsupported resource, got %v", admissionRequest.Resource)
+	}
+	close(domainsToCheck)
+
+	// fetching service data from service central
+	serviceName := getServiceNameFromNamespace(admissionRequest.Namespace)
+	serviceCentralData, err := getServiceData(ctx, scClient, serviceName)
+	if err != nil {
+		return nil, errors.Errorf("error fetching service central data for serviceName %q", serviceName)
+	}
+
+	// types to store parallel domain owner fetches
+	type domainOwnership struct {
+		requestedDomain         string
+		existingDomainAliasInfo *microsserver.AliasInfo
+		err                     error
+	}
+	domainsOwnership := make([]domainOwnership, cap(domainsToCheck))
+
+	// worker groups to parallel fetch requestedDomain owners
+	numWorkers := 5
+	wg := sync.WaitGroup{}
+	wg.Add(numWorkers)
+	var mutex = sync.Mutex{}
+	for i := 0; i < numWorkers; i++ {
+		go func() {
+			defer wg.Done()
+			for domain := range domainsToCheck {
+				aliasInfo, err := microsServerClient.GetAlias(ctx, domain)
+				if aliasInfo != nil {
```

Comment:
Here I'm inside a go func, so I can't make the main func return error. What I'm doing instead is stacking both responses (aliasInfo and err) to the domainsOwnership slice, and them checking errors in the main process later, after wg.Wait.

Is there a better way to do that?

Reply:
ok, I understood better now. We always have to check error first, before dealing with other returns. Anyway, this whole block is gone, because the checks where here just for debugging logs, and the whole block has been removed now.